### DISCUSSION TITLE
GraphQL - Is it possible to edit the schema ?

### DISCUSSION THREAD

**Discussion by Pilpin** [quarkusio/quarkus#34733] [Q&A]
Hello,

I would like to edit the schema so I tried using a function like `public GraphQLSchema.Builder editSchema(@Observes GraphQLSchema.Builder builder)` but as far as I understand I can only add things to the builder, I can not modify the schema.

I tried something like this
```
public GraphQLSchema.Builder edit(@Observes GraphQLSchema.Builder builder) {
    GraphQLSchema _schema = SchemaTransformer.transformSchema(builder.build(), new GraphQLTypeVisitorStub() {
        @Override
        public TraversalControl visitGraphQLObjectType(
                GraphQLObjectType node,
                TraverserContext<GraphQLSchemaElement> context
        ) {
            if (condition) {
                System.out.println(node);
                return changeNode(context, node.transform(b -> {
                    b.doSomething();
                }));
            }

            return super.visitGraphQLObjectType(node, context);
        }
    });

    return GraphQLSchema.newSchema(_schema);
}
```

The `System.out.println(node);` gets executed but the schema at url/q/dev-ui/io.quarkus.quarkus-smallrye-graphql/graphql-schema doesn't change.

Am I missing something ? Is this even possible ?

Thanks.

[a comment by an automated account is not shown]

**Comment by phillip-kruger** [comment]
What do you want to add to the schema ? There might be easier ways to do it. What you have above should work. Can you share a reproducer ?

**Comment by Pilpin** [comment]
I had java interfaces that behaved as graphQL Types in the schema thanks to the `@Type` annotation
```
@Type("MyType")
public interface MyInterface {
...
}
```

The problem is now I need to move theses interfaces to a library in which I won't be able to use the `@Type` annotation, so I thought I could try and change this during graph building.

Is there a way to "annotate" classes and interfaces in libraries ?
I've found this https://github.com/smallrye/smallrye-graphql/issues/316 but I'm not sure I'll be able to do what I want.

I'll try and make a reproducer today.

**Comment by Pilpin** [comment]
Here's the reproducer https://github.com/Pilpin/quarkus-graphql-schema-bug

I don't think the builder returned is used.

**Reply by phillip-kruger** [reply]
I'll have a look at your reproducer, but so far what I can see is you try to change the Country to be another object in GraphQL ? Have you looked at Adapters ? That might be what you want.

**Reply by Pilpin** [reply]
In the reproducer I'm trying to add a field to the country object, and the logs seem to show that it works, up until the observing function returns.
The `Adapter` class and `@AdaptWith` annotation could be a solution but I was trying not to create another class

**Reply by phillip-kruger** [reply]
When you return the original builder, with some modification. does that work ?

**Reply by Pilpin** [reply]
Yes it does. But as far as I know I can't edit the schema this way, I can only add objects to it.

**Reply by phillip-kruger** [reply]
We probably need a new feature in smallrye graphql to allow adding a Transformer. Please create a feature request in Smallrye GraphQL

**Reply by cescoffier** [reply]
@phillip-kruger Do you know if the issue was created?

**Reply by phillip-kruger** [reply]
I don't think so but I'll have a closer look tomorrow

**Comment by nirue** [comment]
Any news about this? I would like to use this feature to replace description for field definitions.

**Comment by nirue** [comment]
public GraphQL.Builder beforeSchemaBuild(@Observes GraphQL.Builder builder) {
        var schema = builder.build().getGraphQLSchema();
        return builder.schema(SchemaTransformer.transformSchema(schema, getTypeVisitor()));
    }

    private GraphQLTypeVisitor getTypeVisitor() {
        return new GraphQLTypeVisitorStub() { ...

There is an undocumented observer method that fits well and satisfy my approach.

**Reply by phillip-kruger** [reply]
Yes. That is the way to do it. ^^^
